### ISSUE TITLE
sozi-export command won't install in Node.js

### ISSUE THREAD

**Issue by wondermoi** [senshu/Sozi-export#13]
Hi Guillaume,
The title says it all. Here is the log file. I hope you will find all the information you need in it. If not, just ask me ! 
[2017-06-14T07_51_41_411Z-debug - Bloc-notes.pdf](https://github.com/senshu/Sozi-export/files/1073985/2017-06-14T07_51_41_411Z-debug.-.Bloc-notes.pdf)

I'm really looking forward to be able to export sozi presentations to pdf and video. Many thanks by the way for the great Sozi app !!

Regards,

**Comment by leycec** [comment]
Sadly, @wondermoi has the right of it: **`sozi-export` is currently uninstallable on *all* supported platforms.**

## SlimerJS: DIE, DIE, DIE

Why? Because [SlimerJS](https://github.com/gmetais/grunt-devperf/issues/26), a hard dependency of `sozi-export`, is fundamentally broken and has been for several years. Specifically, SlimerJS installation fails with the following erroneous 301 HTTP redirect:

```
sudo npm install -g sozi-export
npm WARN deprecated slimerjs@0.9.6: Switch to 0.10.2 when it's released
npm WARN deprecated npmconf@2.1.2: this package has been reintegrated into npm and is now out of date with respect to npm
/usr/local/bin/sozi-to-pptx -> /usr/local/lib/node_modules/sozi-export/cli/sozi-to-pptx.js
/usr/local/bin/sozi-to-video -> /usr/local/lib/node_modules/sozi-export/cli/sozi-to-video.js
/usr/local/bin/sozi-to-pdf -> /usr/local/lib/node_modules/sozi-export/cli/sozi-to-pdf.js

> slimerjs@0.9.6 install /usr/local/lib/node_modules/sozi-export/node_modules/slimerjs
> node install.js

Downloading http://download.slimerjs.org/releases/0.9.6/slimerjs-0.9.6-linux-x86_64.tar.bz2
Saving to /usr/local/lib/node_modules/sozi-export/node_modules/slimerjs/slimerjs/slimerjs-0.9.6-linux-x86_64.tar.bz2
Receiving...
Error requesting archive.
Status: 301
Request options: {
  "protocol": "http:",
  "slashes": true,
  "auth": null,
  "host": "download.slimerjs.org",
  "port": null,
  "hostname": "download.slimerjs.org",
  "hash": null,
  "search": null,
  "query": null,
  "pathname": "/releases/0.9.6/slimerjs-0.9.6-linux-x86_64.tar.bz2",
  "path": "/releases/0.9.6/slimerjs-0.9.6-linux-x86_64.tar.bz2",
  "href": "http://download.slimerjs.org/releases/0.9.6/slimerjs-0.9.6-linux-x86_64.tar.bz2"
}
Response headers: {
  "server": "nginx/1.10.3",
  "date": "Thu, 27 Jul 2017 21:29:24 GMT",
  "content-type": "text/html",
  "content-length": "185",
  "connection": "close",
  "location": "https://download.slimerjs.org/releases/0.9.6/slimerjs-0.9.6-linux-x86_64.tar.bz2"
}
Make sure your network and proxy settings are correct.
/usr/local/lib
└── (empty)

npm ERR! Linux 4.4.0-87-generic
npm ERR! argv "/usr/bin/nodejs" "/usr/bin/npm" "install" "-g" "sozi-export"
npm ERR! node v4.2.6
npm ERR! npm  v3.5.2
npm ERR! code ELIFECYCLE

npm ERR! slimerjs@0.9.6 install: `node install.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the slimerjs@0.9.6 install script 'node install.js'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the slimerjs package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node install.js
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs slimerjs
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls slimerjs
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /home/pietakio/npm-debug.log
npm ERR! code 1
```

[gmetais](https://github.com/gmetais), the principal author of SlimerJS, has [officially acknowledged](https://github.com/gmetais/grunt-devperf/issues/26#issuecomment-302436600) this unresolved issue and admits that it's unlikely to ever be resolved:

> I'm sorry but **I can't maintain this tool anymore.** You might want to take a look at https://github.com/gmetais/YellowLabTools or https://github.com/gmetais/grunt-yellowlabtools.

# SlimerJS Is Dead, But We Still Live

The simplest short-term fix for this critical installation issue would be to embed the last stable release of SlimerJS (*0.9.6?*) directly within the `sozi-export` source tree.

Since SlimerJS is no longer actively maintained, however, a preferable long-term fix would be to refactor `sozi-export` to depend upon an actively maintained alternative. I... have no idea what that might actually be. [YellowLabTools](https://github.com/gmetais/YellowLabTools), maybe?

In short, this *really* needs to be resolved as quickly as feasible.

**Comment by aumouvantsillage** [comment]
> I... have no idea what that might actually be. YellowLabTools, maybe?

Actually, YellowLabTools is based on PhantomJS, so it would be more relevant to use PhantomJS directly.

For the record, I chose to use SlimerJS because PhantomJS did not support SVG. Now, according to #10, this is no longer true.

**Comment by aumouvantsillage** [comment]
> gmetais, the principal author of SlimerJS, has officially acknowledged this unresolved issue and admits that it's unlikely to ever be resolved:

As I understand it, gmetais commented about the tool `grunt-devperf`, that also depends on SlimerJS, and that has the same issue as `sozi-export`. gmetais pointed YellowLabTools as an alternative to `grunt-devperf`.

At https://github.com/laurentj/slimerjs there is no indication that SlimerJS is dead.

It seems that using a more recent version of SlimerJS could solve this issue, as suggested by the error messages that you pasted:

> npm WARN deprecated slimerjs@0.9.6: Switch to 0.10.2 when it's released

**Comment by aumouvantsillage** [comment]
> It seems that using a more recent version of SlimerJS could solve this issue

SlimerJS 0.10 requires Firefox to be installed separately. The latest version of SlimerJs (0.10.3) does not support Firefox 53.

**Comment by aumouvantsillage** [comment]
I have just published a new version based on PhantomJS.
Please try to install `sozi-export` 17.07.30 and tell me if it works as you expect.

**Comment by wondermoi** [comment]
Unfortunately, my latest attempts failed...

> Your environment has been set up for using Node.js 8.1.1 (x64) and npm.
> 
> C:\Users\Maman>npm install -g sozi-export
> C:\Users\Maman\AppData\Roaming\npm\sozi-to-pptx -> C:\Users\Maman\AppData\Roaming\npm\node_modules\sozi-export\cli\sozi-to-pptx.js
> C:\Users\Maman\AppData\Roaming\npm\sozi-to-video -> C:\Users\Maman\AppData\Roaming\npm\node_modules\sozi-export\cli\sozi-to-video.js
> C:\Users\Maman\AppData\Roaming\npm\sozi-to-pdf -> C:\Users\Maman\AppData\Roaming\npm\node_modules\sozi-export\cli\sozi-to-pdf.js
> 
> > phantomjs-prebuilt@2.1.14 install C:\Users\Maman\AppData\Roaming\npm\node_modules\sozi-export\node_modules\phantomjs-prebuilt
> > node install.js
> 
> PhantomJS not found on PATH
> Downloading https://github.com/Medium/phantomjs/releases/download/v2.1.1/phantomjs-2.1.1-windows.zip
> Saving to C:\Users\Maman\AppData\Local\Temp\phantomjs\phantomjs-2.1.1-windows.zip
> Receiving...
>   [========================================] 100%
> Received 17767K total.
> Extracting zip contents
> Removing C:\Users\Maman\AppData\Roaming\npm\node_modules\sozi-export\node_modules\phantomjs-prebuilt\lib\phantom
> Copying extracted folder C:\Users\Maman\AppData\Local\Temp\phantomjs\phantomjs-2.1.1-windows.zip-extract-1501482611599\phantomjs-2.1.1-windows -> C:\Users\Maman\AppData\Roaming\npm\node_modules\sozi-export\node_modules\phantomjs-prebuilt\lib\phantom
> Writing location.js file
> Done. Phantomjs binary available at C:\Users\Maman\AppData\Roaming\npm\node_modules\sozi-export\node_modules\phantomjs-prebuilt\lib\phantom\bin\phantomjs.exe
> + sozi-export@17.7.30
> added 134 packages in 60.305s
> 
> C:\Users\Maman>sozi-to-video  -h
> 
>   Usage: sozi-to-video [options]
> 
> 
>   Options:
> 
>     -o, --output <file>      Output file
>     -W, --width <number>     Video width, in pixels (defaults to 1024)
>     -H, --height <number>    Video height, in pixels (defaults to 768)
>     -b, --bit-rate <number>  Video bit rate, in bits/second (defaults to 2M)
>     -h, --help               output usage information
> 
> C:\Users\Maman>sozi-to-video presentation.sozi.html
> Converting presentation.sozi.html to presentation.sozi.ogv
> events.js:182
>       throw er; // Unhandled 'error' event
>       ^
> 
> Error: spawn avconv ENOENT
>     at exports._errnoException (util.js:1016:11)
>     at Process.ChildProcess._handle.onexit (internal/child_process.js:189:19)
>     at onErrorNT (internal/child_process.js:366:16)
>     at _combinedTickCallback (internal/process/next_tick.js:102:11)
>     at process._tickCallback (internal/process/next_tick.js:161:9)
> 
> C:\Users\Maman>sozi-to-video test_prog.sozi.html
> Converting test_prog.sozi.html to test_prog.sozi.ogv
> sozi-to-video.js> Exporting frame: 1
> sozi-to-video.js> Exporting frame: 2
> sozi-to-video.js> Exporting frame: 3
> sozi-to-video.js> Exporting frame: 4
> events.js:182
>       throw er; // Unhandled 'error' event
>       ^
> 
> Error: spawn avconv ENOENT
>     at exports._errnoException (util.js:1016:11)
>     at Process.ChildProcess._handle.onexit (internal/child_process.js:189:19)
>     at onErrorNT (internal/child_process.js:366:16)
>     at _combinedTickCallback (internal/process/next_tick.js:102:11)
>     at process._tickCallback (internal/process/next_tick.js:161:9)
> 
> C:\Users\Maman>

Did I miss anything ?

**Comment by aumouvantsillage** [comment]
> Did I miss anything ?

Have you installed libav?

**Comment by wondermoi** [comment]
> Have you installed libav?

Yes I did, through ConvertHelper 3.2, since I operate from Win10.  Any clue ?

**Comment by aumouvantsillage** [comment]
I don't know how ConvertHelper is installed.
Can you check that `avconv` can be run from the command line.
If not, maybe you should try to install the official build from http://builds.libav.org/windows/release-gpl/
I have never done that, so I'm not sure I can help you on this.

**Comment by wondermoi** [comment]
I can run `avconv` from the command line but only within the folder I install the official _libav_ build...  Now I have no clue how `sozi-export` can access it without knowing this very path... Is there anything to do in the registry ? As you can imagine, I am no coder and I feel pretty lost ! 
Anyway, I very much appreciate your trying to help me. Thx

**Comment by aumouvantsillage** [comment]
> Now I have no clue how sozi-export can access it without knowing this very path

There should be an environment variable named `PATH`.
To execute `avconv` from any folder, the `PATH` variable must include the folder where `avconv` is installed.

You can find more info about this here: [What are Environment Variables?](https://superuser.com/a/284351)

**Comment by wondermoi** [comment]
`PATH` environment variable set !
`Sozi-export` now access `avconv` on the system and here is the new error I get -> 

> Your environment has been set up for using Node.js 8.1.4 (x64) and npm.
> 
> C:\Users\Maman\Documents\[...]\Test Sozi>sozi-to-video test_prog_sozi.html
> Converting test_prog_sozi.html to test_prog_sozi.ogv
> avconv version 11.3, Copyright (c) 2000-2014 the Libav developers
>   built on Mar  9 2015 02:12:20 with gcc 4.8.3 (Gentoo 4.8.3 p1.1, pie-0.5.9)
> **C:\Users\Maman\AppData\Local\Temp\tmp-9340s6EQL7BOrj4Y\%d.png: No such file or directory**

I can feel I'm close to successfully convert a Sozi presentation to a video file...

**Comment by aumouvantsillage** [comment]
> I can feel I'm close to successfully convert a Sozi presentation to a video file...

Yes. It means that `sozi-export` has successfully generated a set of image files.
Now it fails to create a video from these image files.

> C:\Users\Maman\AppData\Local\Temp\tmp-9340s6EQL7BOrj4Y%d.png: No such file or directory

There is something wrong with the file names given to `avconv`. It should read `C:\Users\Maman\AppData\Local\Temp\tmp-9340s6EQL7BOrj4Y\%d.png` (with a backslash before the `%` character).

This is a bug in `sozi-export` or node.js. Since it worked correctly on Linux, this is probably a Windows-specific problem.

A workaround would be to call `avconv` manually with the correct path. However, `sozi-export` deletes all temporary files before terminating, so it will probably not be so easy without modifying the source code.

I'll get back to you when I have a solution.

**Comment by aumouvantsillage** [comment]
I have published a new version: 17.08.01.
It attempts to fix the issue with image file paths when converting to video on Windows.

Run the following command to upgrade:

```
npm update -g sozi-export
```

Then you can retry to run `sozi-to-video`. Please tell me if it works.

**Comment by wondermoi** [comment]
One step further... here is the trace I get ->

> C:\Users\Maman\[...]Test Sozi>sozi-to-video test_prog.sozi.html
> Converting test_prog.sozi.html to test_prog.sozi.ogv
> sozi-to-video.js> Exporting frame: 1
> sozi-to-video.js> Exporting frame: 2
> sozi-to-video.js> Exporting frame: 3
> sozi-to-video.js> Exporting frame: 4
> avconv version 11.3, Copyright (c) 2000-2014 the Libav developers
>   built on Mar  9 2015 02:12:20 with gcc 4.8.3 (Gentoo 4.8.3 p1.1, pie-0.5.9)
> Input #0, image2, from 'C:\Users\Maman\AppData\Local\Temp\tmp-6012cL9ahVfJtdI3\img%d.png':
>   Duration: 00:00:03.00, start: 0.000000, bitrate: N/A
>     Stream #0.0: Video: png, bgra, 1024x784, 50 tbn
> Output #0, ogg, to 'test_prog.sozi.ogv':
>     Stream #0.0: Video: [0][0][0][0] / 0x0000, q=2-31, 200 kb/s
> Stream mapping:
>   Stream #0:0 -> #0:0 (png (native) -> ? (?))
> **Automatic encoder selection failed for output stream #0:0. Default encoder for format ogg is probably disabled. Please choose an encoder manually.**

Using VLC, I thought it would be OK for any format encoding, but it seems it's not. I then install an ogv encoder,` libtheora-1.1.1` found on [https://wiki.xiph.org/TheoraEncoders](https://wiki.xiph.org/TheoraEncoders), but I don't know how to use it...

**Comment by aumouvantsillage** [comment]
I don't know either.

If you don't care about the target video format, you can try to run `avconv -formats`. It will tell you which formats are available for encoding (those preceded by letter 'E').
When running `sozi-to-video`, set the target file extension according to the desired video format.

**Comment by wondermoi** [comment]
Yesss ! It works just fine with .mp4 encoding format.
Thank you so much for your precious help, I've learned a lot with you.
